### PR TITLE
feat(FN-6880): graph-native optional steps + add-on subgraphs (and FN-6879 node help)

### PR DESCRIPTION
**Pull request by gsxdsm**
## Summary

Two workflow-editor improvements (the branch-group work this branch is named for is already on `main`):

### FN-6880 — Graph-native optional steps + add-ons as subgraphs
Optional steps move from an execution-inert *declaration* (`optionalSteps: [{templateId}]` run through a hidden `workflow-step` seam) to a real graph construct:

- **`optional-group` container node** (mirrors `foreach`/`loop`) holding a `template:{nodes,edges}` subgraph. The graph executor runs it **once** when the group is enabled for a task and **bypasses** it (passes through, runs no template node) when disabled. (U1 IR+validation, U2 executor)
- **Enable state reuses the per-task `enabledWorkflowSteps` facet** + a workflow-level `defaultOn`, keyed by the group node id; new tasks seed from `defaultOn`. (U3)
- **Editor authoring**: the container is a registered React-Flow group node with a `defaultOn` toggle; round-trips through `flowToIr`/`irToFlow`. (U4)
- **All 7 built-in add-ons** (documentation-review, qa-check, security-audit, performance-review, accessibility-check, browser-verification, frontend-ux-design) are insertable from the palette as a node **or** wrapped in an optional-group. (U5)
- **Built-ins migrated**: coding + stepwise-coding express `browser-verification` as an optional-group (default OFF); coding is now interpreter-deferred. Parity oracles updated. (U6)
- **Legacy declaration surface retired** (U7a): the `WorkflowOptionalStep` type + `optionalSteps` IR field + `validateOptionalSteps`, and the editor's declaration authoring UI (`WorkflowOptionalStepsPanel`, `optionalStepsOf`, `flowToIr` threading). A legacy persisted `optionalSteps` key is tolerated at parse. The per-task toggle surfaces (dropdown, inline card, modal, Workflow tab) are unchanged — they consume the distinct `ResolvedWorkflowOptionalStep`.

### FN-6879 — Workflow node Help in the detail pane
Every node kind (including engine-managed graph-only nodes: merge gate, branch-group integration/promotion, PR/recovery) gets an in-editor Help section: what it does, how to configure it, inputs/outputs/edges.

## Code review
4 reviewers (correctness, testing, API-contract, reliability) over the feature diff. One **P1** found and fixed: enabling a built-in optional-group whose node id collides with a `WORKFLOW_STEP_TEMPLATES` id was silently bypassed (the toggle id was remapped to a materialized step row id the executor never matched) — fixed by passing optional-group ids through enable resolution untouched, with colliding-id regression tests on the create + update paths. Triaged P3s (description/icon drop, switch-time seeding, selection-row id mix) are benign/deferred.

## Verification
- Lint clean · `tsc --noEmit` 0 errors in core/engine/dashboard · `pnpm build` succeeds · `pnpm test:gate` green (engine-core 649, ci-shape 58).
- Feature suites: core IR/validation/resolver/seeding, engine run-once/bypass + two-task divergence + built-in execution, dashboard editor round-trip + palette + node-help.

## Deferred (documented follow-ups)
- **`workflow-step` seam infrastructure removal** — the seam is a shared `WorkflowSeam` union member woven through ~9 engine runtime files (`runtime-primitives`, `step-session-executor`, `workflow-node-handlers`, compiler seam-anchor, `runWorkflowSteps`). Now orphaned (no built-in uses it) but inert; excising it is its own unit. See the plan's Deferred section.
- Minor: optional-group `description`/`icon` on the toggle UI; `defaultOn` seeding on post-create workflow switch.

## Not run
- Browser/visual testing was not run in this environment (worktree-bundle dashboard setup). UI is covered by jsdom render tests including the node-type registration test (the exact stale-bundle/unregistered-type failure mode the team's learnings flag). Recommend a real-browser pass before merge.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://stagereview.app/Runfusion/Fusion/pull/1712">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

## Release Notes

**New Features**
- Optional-group container nodes now power optional workflow behavior (run once per enabled task; bypass when disabled).
- Added a Help section to the workflow node inspector, including “Engine-managed” read-only nodes.
- Added “insert as optional group” to wrap palette templates quickly.

**Improvements**
- Built-in coding/stepwise workflows now use optional-group for browser verification.
- Task enablement supports optional-group `defaultOn` seeding.

**Bug Fixes**
- Fixed cases where optional-group toggles could be silently bypassed when ids collide.

**Changes (Breaking)**
- Retired the legacy optional-step model and its authoring UI; migrate to optional-group nodes.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 12 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Define optional-group IR and validation](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/1) |
| 2 | [Implement optional-group execution logic](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/2) |
| 3 | [Resolve per-task enablement from group nodes](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/3) |
| 4 | [Fix toggle-id collision in workflow store](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/4) |
| 5 | [Map optional-group nodes in the editor](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/5) |
| 6 | [Register optional-group node types and UI](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/6) |
| 7 | [Enable optional-group authoring in the editor](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/7) |
| 8 | [Project add-ons as optional-group subgraphs](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/8) |
| 9 | [Migrate built-in workflows to optional-group](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/9) |
| 10 | [Update built-in parity oracles and tests](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/10) |
| 11 | [Retire legacy optional-step declarations](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/11) |
| 12 | [Add node help to the detail pane](https://stagereview.app/Runfusion/Fusion/pull/1712/chapters/12) |

<a href="https://stagereview.app/Runfusion/Fusion/pull/1712">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 37b2cb38aceb0645cd768addd2fca1c89565dd48 on Jun 24, 2026 12:17am UTC.</sub>

**Comment by gsxdsm** [comment]
## ✅ Real-browser verification (agent-browser)

Ran the dashboard from the worktree dist (`--dev`, engine off, port 4101, `FUSION_CLIENT_DIR` → fresh `packages/dashboard/dist/client`) per the team's worktree-testing learning. Verified the built-in **Coding** workflow in the node editor:

- **Node-type registration (the stale-bundle failure mode):** `optional-group` renders with its registered testid; **0** `react-flow__node-default` nodes; 20 nodes total — the container is a real registered group, not the default fallback.
- **Container:** renders as **"Browser Verification"** with a **"default off"** badge.
- **Inspector:** heading **"Optional group"**, the **"Enabled by default for new tasks"** (`defaultOn`) toggle, and the **Help** section ("Runs the steps inside this group once when the task enables it … skips them when disabled").
- **Add-on palette (U5):** all **7** add-ons present with **both** insert variants — `wf-tpl-step-<id>` (as node) and `wf-tpl-step-<id>-optional-group` ("as optional group"): accessibility-check, browser-verification, documentation-review, frontend-ux-design, performance-review, qa-check, security-audit.
- **Node Help (FN-6879):** confirmed rendering in the same pass.

This supersedes the "browser testing not run" caveat in the description.

File: .changeset/retire-optional-steps-declaration.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@runfusion/fusion": patch
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply .changeset/retire-optional-steps-declaration.md]: > Breaking-change bump should be `major`, not `patch`

Addressed: bumped the changeset to `major` and called out the removed `WorkflowOptionalStep` export. (e4a810e9b)

File: packages/engine/src/workflow-graph-loop.ts [hunk 2]
```diff
@@ -208,3 +208,87 @@ export async function runLoop(
...
+
+    if (lastResult.outcome === "failure") {
+      // Publish accumulated template context, then surface the failure as the
+      // group's outcome so its failure/outcome: edges route.
+      Object.assign(env.context, groupContext);
+      return { outcome: "failure", value: lastResult.value, visitedNodeIds };
+    }
+
+    const edges: WorkflowIrEdge[] = outgoing.get(current.id) ?? [];
+    const matching: WorkflowIrEdge[] = edges.filter((edge: WorkflowIrEdge) =>
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/workflow-graph-loop.ts:285]: > Failure-condition edges inside optional-group templates silently never execute

Addressed: `validateOptionalGroup` now rejects `condition: "failure"` edges inside the template. The single-pass walk surfaces a template-node failure as the group's outcome (routing the group's *outer* failure edge), so an internal failure edge is now a typed authoring error instead of a dead edge. (e4a810e9b)

File: docs/plans/2026-06-21-002-feat-workflow-optional-group-subgraphs-plan.md [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+- **Editor:** `packages/dashboard/app/components/workflow-flow-mapping.ts:46-82,266-269,316-403,431-550`
+  (group children, `groupTemplateConfigOf`), `:1110-1219` (`insertFragment` subgraph insertion);
+  `WorkflowNodeEditor.tsx:985-1010` (palette: fragments/steps/plugins, sourced from
+  `/api/workflow-step-templates`), `:1425-1453` (insert handlers); `nodes/WorkflowNodeTypes.tsx` (node-type
+  registry to extend).
+- **Compiler:** `packages/core/src/workflow-compiler.ts:45,165-196` (seam anchors + canonical order).
+- **Built-ins:** `packages/core/src/builtin-coding-workflow-ir.ts:69,123`,
+  `builtin-stepwise-coding-workflow-ir.ts:132`.
+- **Prior art:** `docs/plans/2026-06-20-001-feat-workflow-optional-steps-node-editor-modal-plan.md` (the
+  declaration-based system this replaces); institutional learnings cited inline under Risks.
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply docs/plans/2026-06-21-002-feat-workflow-optional-group-subgraphs-plan.md:626]: > Add documentation of the P1 fix (U7a: optional-group id collision handling) as a separate unit.

The P1 fix is captured in the commit history (`a38acaf86`) and as a dedicated learning at `docs/solutions/logic-errors/optional-group-toggle-id-remapped-by-step-materializer.md`. The plan is a pre-execution decision artifact (delivered-unit progress is derived from git, not retrofitted into the plan body); I updated its scope note to record the U1–U6 + U7a cohort rather than adding post-hoc units.

File: docs/plans/2026-06-21-002-feat-workflow-optional-group-subgraphs-plan.md [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+  rather than the per-task toggle) — single-level, per-task-toggle only for now.
+- **Plugin-contributed add-ons as optional-group presets** beyond inserting them as flat nodes.
+- The prior plan's deferred **unified per-task workflow-facet override** abstraction (optional steps +
+  auto-merge + column-agent overrides) — strong `/ce-compound` candidate once this lands.
+
+**Out of scope:**
+- New persistence/migrations. Reuses `tasks.enabledWorkflowSteps`; `optional-group` is just a node kind.
+- Rewriting `WorkflowStepTemplate` into a nodes+edges shape (KTD-5 keeps it flat).
+- Changing unrelated execution (merge lifecycle, branch groups, PR nodes).
+
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply docs/plans/2026-06-21-002-feat-workflow-optional-group-subgraphs-plan.md:567]: > Verify Scope & Deferred sections align with actual PR cohort delivery.

Addressed: the Deferred section was stale — U7a (declaration-model retirement) shipped in this PR. Replaced it with a "Delivered cohort (this PR) vs. Deferred" note recording U1–U6 + U7a + the review P1 fix as delivered, leaving only the `workflow-step` seam-infrastructure removal deferred. (e4a810e9b)

File: packages/core/src/workflow-ir.ts [hunk 5]
```diff
@@ -1381,12 +1465,17 @@ export function downgradeIrToV1IfPure(ir: WorkflowIr): WorkflowIr {
...
+  // `optionalSteps` is no longer a typed IR field (retired declaration model), but
+  // a legacy v2 row may still carry the key. Read it via an untyped cast so such a
+  // row is still treated as v2 (kept on v2, never silently downgraded).
+  const legacyOptionalSteps = (ir as { optionalSteps?: unknown[] }).optionalSteps;
   if (
     (ir.artifacts && ir.artifacts.length > 0) ||
     (ir.fields && ir.fields.length > 0) ||
     (ir.settings && ir.settings.length > 0) ||
-    (ir.optionalSteps && ir.optionalSteps.length > 0)
+    (Array.isArray(legacyOptionalSteps) && legacyOptionalSteps.length > 0)
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/workflow-ir.ts]: > Keep any legacy `optionalSteps` key as v2-only during downgrade checks.

Addressed: the downgrade heuristic now treats the mere presence of a legacy `optionalSteps` key (including `[]`) as a v2 signal, so an `optionalSteps: []` row is no longer silently downgraded to v1. (e4a810e9b)

File: packages/core/src/workflow-optional-steps.ts [hunk 6]
```diff
@@ -10,37 +14,72 @@ export interface ResolvedWorkflowOptionalStep {
...
+ * toggle plus their seed default.
  *
- * The declaration is intentionally execution-inert: it only advertises which
- * template-backed workflow steps a task may toggle into `enabledWorkflowSteps`.
- * Unknown template ids are skipped so stale/custom declarations never render
- * blank UI rows or break workflow loading.
+ * Source: v2 `ir.nodes` where `kind === "optional-group"` (NOT the legacy
+ * `ir.optionalSteps` declaration). Non-v2 graphs and graphs without any
+ * optional-group node resolve to `[]`. Malformed group configs are skipped so a
+ * stale/partial node never renders a blank UI row or breaks workflow loading.
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/workflow-optional-steps.ts]: > Docblock behavior is out of sync with the resolver implementation.

Addressed: corrected the docblock — a group with a missing/partial config resolves to a usable entry (name falls back to the node id, defaultOn to false) rather than being skipped. (e4a810e9b)

File: packages/dashboard/app/components/__tests__/workflow-flow-mapping.test.ts [hunk 7]
```diff
@@ -1338,6 +1436,38 @@ describe("insertFragment", () => {
...
+    const allIds = second.nodes.map((n) => n.id);
+    expect(new Set(allIds).size).toBe(allIds.length);
+
+    // Round-trip: the group carries defaultOn + a single-node template.
+    const { ir: out } = flowToIr("wf", second.nodes, second.edges);
+    const og = out.nodes.find((n) => n.kind === "optional-group")!;
+    expect(og.config?.defaultOn).toBe(true);
+    const template = (og.config as { template?: { nodes: { config?: Record<string, unknown> }[] } }).template;
+    expect(template?.nodes).toHaveLength(1);
+    expect(template?.nodes[0].config?.name).toBe("Security Audit");
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/__tests__/workflow-flow-mapping.test.ts]: > Assert the round-trip invariant for both inserted groups.

Addressed: the test now iterates all optional-group nodes (both must carry `defaultOn` + a single-node template) and also asserts the round-trip serializes as v2. (e4a810e9b)

File: packages/dashboard/app/components/__tests__/WorkflowNodeEditor.test.tsx [hunk 8]
```diff
@@ -2798,6 +2907,123 @@ describe("WorkflowNodeEditor — U9 palette Templates section", () => {
...
+    );
+
+    fireEvent.click(screen.getByText("Save").closest("button")!);
+    await waitFor(() => expect(updateWorkflow).toHaveBeenCalled());
+    const [, updates] = vi.mocked(updateWorkflow).mock.calls[0];
+    const ir = (updates as { ir: { nodes: { id: string; kind: string }[] } }).ir;
+    const groupIds = ir.nodes.filter((n) => n.kind === "optional-group").map((n) => n.id);
+    expect(groupIds).toHaveLength(2);
+    expect(new Set(groupIds).size).toBe(2);
+  });
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/__tests__/WorkflowNodeEditor.test.tsx:3266]: > Broaden regression coverage beyond the populated happy path.

The empty / undefined / duplicate-template states are enforced at the validation layer and covered in `packages/core/src/__tests__/workflow-ir-optional-group.test.ts` (empty-template, missing-template, duplicate-id, plus the new failure-edge case). The editor test covers the populated authoring round-trip + cascade-delete; a full editor-layer surface matrix would duplicate the validation guarantees, so coverage is kept where it is strongest.

File: packages/dashboard/app/components/WorkflowNodeEditor.tsx [hunk 9]
```diff
@@ -1355,16 +1342,19 @@ function InnerEditor({
...
+        // FNXC:WorkflowOptionalGroup 2026-06-21-11:30: An optional-group is authored exactly like a foreach/loop region — drop nodes inside; the subgraph runs once when the task enables the group.
         const childId = foreachChildFlowId(id, newNodeId());
         const childLabel =
           kind === "foreach"
             ? t("workflowNodes.stepExecuteLabel", "Step execute")
-            : t("workflowNodes.loopStepLabel", "Loop step");
+            : kind === "optional-group"
+              ? t("workflowNodes.optionalGroupStepLabel", "Optional step")
+              : t("workflowNodes.loopStepLabel", "Loop step");
         const childConfig = kind === "foreach" ? { seam: "step-execute" } : { prompt: "" };
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/WorkflowNodeEditor.tsx:1458]: > Make optional-group presence force v2 serialization.

Addressed: `flowToIr`'s v2 signal now includes container/group nodes (foreach/loop/optional-group). An inserted optional-group on an otherwise-plain workflow now serializes as v2 instead of an invalid v1 that would fail parse. Added a round-trip v2 assertion. (e4a810e9b)

File: packages/engine/src/__tests__/builtin-coding-browser-verification-group.test.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      return { outcome: "success" };
+    });
+    const executor = new WorkflowGraphExecutor({ handlers: { prompt } });
+
+    const result = await executor.run(codingTask([GROUP_ID]), settingsOn(), BUILTIN_CODING_WORKFLOW_IR);
+
+    expect(result.context[`node:${GROUP_ID}:outcome`]).toBe("failure");
+    expect(result.visitedNodeIds).toContain(INNER_STEP_VISITED_ID);
+    expect(result.visitedNodeIds).not.toContain("review");
+  });
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/__tests__/builtin-coding-browser-verification-group.test.ts:97]: > Assert the terminal-node invariant explicitly

`end` is a terminal node the executor does not record in `visitedNodeIds` (the failure run ends at the group's inner step — 5 nodes, no `end`), so `toContain("end")` is not expressible. The failure routing is asserted via the group's `failure` outcome in context plus `review` being unreachable; the group's only two outgoing edges are `success → review` and `failure → end`, so a skipped review proves the failure edge was taken. Clarified the test comment.

File: packages/engine/src/workflow-graph-executor.ts [hunk 11]
```diff
@@ -484,6 +484,50 @@ export class WorkflowGraphExecutor {
...
+            // FNXC:WorkflowOptionalGroup 2026-06-21-16:30: record the group's own
+            // outcome on bypass too (mirrors the enabled path + every other node
+            // kind), so a downstream node reading `node:<id>:outcome` from context
+            // sees "success" rather than undefined — disabled is fully inert, not
+            // just edge-routing-inert.
+            context[`node:${node.id}:outcome`] = "success";
+            return await traverseChildren(node, {
+              outcome: "success",
+              value: "optional-group-bypassed",
+            });
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/workflow-graph-executor.ts]: > Disabled optional-groups can route non-inertly because of the synthetic bypass value.

Addressed: the disabled bypass now returns a plain `{ outcome: "success" }` with no distinguishing value, so an `outcome:*` edge can no longer preempt the success edge in `traverseChildren` — a disabled group routes identically to the node being absent. (e4a810e9b)

**Comment by gsxdsm** [comment]
Addressed the two out-of-diff CodeRabbit comments on the changesets (commit 95ec4bcdb):

> **Document the P1 fix (optional-group id collision handling) in the changeset.**

Added to `workflow-optional-group-subgraphs.md`: "Optional-group enable resolution correctly handles id collisions with add-on template ids, so a group's enable state is not silently bypassed during task creation/update."

> **Reconcile inconsistency … regarding legacy surface retirement.**

Fixed the contradiction — `workflow-optional-group-subgraphs.md` no longer says the legacy declaration surface "remains for back-compat" (U7a retired it; `retire-optional-steps-declaration.md` documents the removal). Only the `workflow-step` seam-infrastructure removal remains a follow-up.
